### PR TITLE
Add result utilities

### PR DESCRIPTION
**Pull request by l4l**
### Description of the Change

Utility functions for result usage:
- <s>has_val</s>
- <s>has_err</s>
- some_val (fixture)
- some_err (fixture)
- and_res
- or_res
- <s>rmap</s>
- <s>r</s>map_err<b>or</b>

And test & doc for them

Replaced with stricter analogue:
`ASSERT_NO_THROW` + `checkValueCase` -> `ASSERT_TRUE` + `val`
`ASSERT_NO_THROW` + `checkErrorCase` -> `ASSERT_TRUE` + `err`

Replaced as more generic:
`checkValueCase` -> `val`
`checkErrorCase` -> `err`

### Benefits

There will be much more readable code that deals with multiple errors, eg this block:
https://github.com/hyperledger/iroha/blob/master/irohad/execution/impl/command_executor.cpp#L89

### Possible Drawbacks 

None

### Tests changed

- ametsuchi_test
- command_validate_execute_test
- result_test
- wsv_query_command_test

### CODE REVIEW

File: libs/common/result.hpp [hunk 1]
```diff
@@ -106,8 +107,73 @@ namespace iroha {
...
+
+      constexpr bool has_val() const {
+        return some_val();
+      }
+
+      constexpr bool has_err() const {
+        return some_err();
+      }
+
+      constexpr auto some_val() const {
```
Comment (grimadas) [inline libs/common/result.hpp]: The naming is not obvious, maybe a comment can help?
Same for all new methods in the file except has_val, has_err

File: libs/common/result.hpp [hunk 2]
```diff
@@ -106,8 +107,73 @@ namespace iroha {
...
+            [](ValueType v) -> boost::optional<ErrorType> { return {}; },
+            [](ErrorType e) { return boost::optional<ErrorType>(e); });
+      }
+
+      /**
+       * @return new_res if this Result contains a value
+       *         otherwise return this
+       */
+      template <typename Value>
+      constexpr Result<Value, E> and_res(Result<Value, E> new_res) const {
```
Comment (grimadas) [inline libs/common/result.hpp]: I'm not sure this is best obvious name, do you have any other candidates?

File: libs/common/result.hpp [hunk 3]
```diff
@@ -106,8 +107,73 @@ namespace iroha {
...
+            [res = new_res](ValueType) { return res; },
+            [](ErrorType err) -> Result<Value, E> { return err; });
+      }
+
+      /**
+       * @return new_res if this Result contains a error
+       *         otherwise return this
+       */
+      template <typename Value>
+      constexpr Result<Value, E> or_res(Result<Value, E> new_res) const {
```
Comment (grimadas) [inline libs/common/result.hpp]: I'm not sure this is best obvious name, do you have any other candidates?

File: test/module/libs/common/result_test.cpp [hunk 4]
```diff
@@ -158,6 +158,146 @@ TEST(ResultTest, ResultVoidError) {
...
+  result.and_res(new_res).match([](Value<int> v) { ASSERT_EQ(4, v.value); },
+                                makeFailCase<Error<void>>(kErrorCaseMessage));
+}
+
+/**
+ * @given Pair of results: first with error, second with value
+ * @when and_res function is invoked
+ * @then result contains the first error
+ */
+TEST(ResultTest, AndResWithErrVal) {
```
Comment (grimadas) [inline test/module/libs/common/result_test.cpp]: What about two errors case ?

File: test/module/libs/common/result_test.cpp [hunk 5]
```diff
@@ -158,6 +158,146 @@ TEST(ResultTest, ResultVoidError) {
...
+  result.or_res(new_res).match([](Value<int> v) { ASSERT_EQ(5, v.value); },
+                               makeFailCase<Error<void>>(kErrorCaseMessage));
+}
+
+/**
+ * @given Pair of results: first with error, second with value
+ * @when or_res function is invoked
+ * @then result contains the second value
+ */
+TEST(ResultTest, OrResWithErrVal) {
```
Comment (grimadas) [inline test/module/libs/common/result_test.cpp]: Also add case with Value and Error

File: libs/common/result.hpp [hunk 6]
```diff
@@ -106,8 +107,73 @@ namespace iroha {
...
+        return visit_in_place(
+            *this,
+            [](ValueType val) -> Result<Value, E> { return val; },
+            [res = new_res](ErrorType) { return res; });
+      }
     };
 
+    template <typename Val1, typename Val2, typename E, typename Fn>
+    auto rmap(Result<Val2, E> res, Fn &&map) {
```
Comment (grimadas) [inline libs/common/result.hpp]: Add comment here pls

File: libs/common/result.hpp [hunk 7]
```diff
@@ -106,8 +107,73 @@ namespace iroha {
...
+      return visit_in_place(
+          res,
+          [map](Value<Val2> val) -> Result<Val1, E> {
+            return Value<Val1>{map(val.value)};
+          },
+          [](Error<E> err) -> Result<Val1, E> { return err; });
+    }
+
+    template <typename Err1, typename Err2, typename V, typename Fn>
+    auto rmap_err(Result<V, Err2> res, Fn &&map) {
```
Comment (grimadas) [inline libs/common/result.hpp]: Add comment here pls

File: libs/common/result.hpp (hunk 6 above)
Comment (nickaleks) [inline libs/common/result.hpp]: How is it different from | operation?

**Comment by nickaleks** [comment]
Could you provide code fragments where these methods could be used? I mean is there an actual need for them. 

When designing result type, we tried to avoid imperative style programming, such as has_val, or has_err, since they allow programmers to skip error handling completely, which is undesired effect.

File: libs/common/result.hpp (hunk 2 above)
Comment (l4l) [reply libs/common/result.hpp]: Not really, in rust it named just `and`, but that is a keyword is c++

File: test/module/libs/common/result_test.cpp (hunk 4 above)
Comment (l4l) [reply test/module/libs/common/result_test.cpp]: Considered them as negligible. Do you think it's important?

File: libs/common/result.hpp (hunk 6 above)
Comment (l4l) [reply libs/common/result.hpp]: Duplicate, indeed

File: test/module/libs/common/result_test.cpp (hunk 4 above)
Comment (grimadas) [reply test/module/libs/common/result_test.cpp]: Will it increase the test coverage?

File: libs/common/result.hpp [hunk 8]
```diff
@@ -106,8 +107,61 @@ namespace iroha {
...
+      constexpr Result<Value, E> or_res(Result<Value, E> new_res) const {
+        return visit_in_place(
+            *this,
+            [](ValueType val) -> Result<Value, E> { return val; },
+            [res = new_res](ErrorType) { return res; });
+      }
     };
 
+    /**
+     * Get a new result with changed error
```
Comment (grimadas) [inline libs/common/result.hpp]: This comment is unclear

File: libs/common/result.hpp [hunk 9]
```diff
@@ -106,8 +107,61 @@ namespace iroha {
...
+     * @param res for operating for
```
Comment (grimadas) [inline libs/common/result.hpp]: 😮 unclear

File: test/module/libs/common/result_test.cpp [hunk 10]
```diff
@@ -158,6 +158,102 @@ TEST(ResultTest, ResultVoidError) {
...
+TEST(ResultTest, MapError) {
+  Result<void, int> result = makeError(5);
+  map_error<int>(result, [](auto i) { return i * 2; })
+      .match(makeFailCase<Value<void>>(kValueCaseMessage),
+             [](Error<int> e) { ASSERT_EQ(10, e.error); });
+}
+
+/**
+ * @given Result with some error and some function
+ * @when rmap function is invoked
```
Comment (grimadas) [inline test/module/libs/common/result_test.cpp]: map_error

**Pull request merged**